File: simulator/src/piwarssim/engine/simulation/MineSweeperStateObject.py
from enum import Enum

from piwarssim.engine.simulation.MovingSimulationObjectWithPositionAndOrientation import MovingSimulationObjectWithPositionAndOrientation
from piwarssim.engine.simulation import SimulationObject


class MineSweeperStateObject(SimulationObject):
    def __init__(self, factory, sim_object_id, sim_object_type):
        super(MineSweeperStateObject, self).__init__(factory, sim_object_id, sim_object_type)
        self._state_bits = 0

    def free(self):
        self._state_bits = 0
        super(MineSweeperStateObject, self).free()

    def set_state_bits(self, state_bits):
        self.changed |= (state_bits != self._state_bits)
        self._state_bits = state_bits

    def get_state_bits(self):
        return self._state_bits

    def set_state_bit(self, x, y, value):
        index = x * 4 + y
        bit = 1 << index
        state_bits = self._state_bits
        if value:
            state_bits |= bit
        else:
            state_bits &= ~bit

        self.changed |= (state_bits != self._state_bits)
        self._state_bits = state_bits

    def get_state_bit(self, x, y):
        index = x * 4 + y
        bit = 1 << index
        return (self._state_bits & bit) != 0

    def serialize(self, full, serializer):
        super(MineSweeperStateObject, self).serialize(full, serializer)
        serializer.serialize_unsigned_short(self._state_bits)

    def deserialize(self, full, serializer):
        super(MineSweeperStateObject, self).deserialize(full, serializer)
        state_bits = serializer.deserialize_unsigned_short()

        self.changed = self.changed or self._state_bits != state_bits

        self._state_bits = state_bits

    def size(self, full):
        return super(MineSweeperStateObject, self).size(full) + 2

    def copy_internal(self, new_object):
        super(MineSweeperStateObject, self).copy_internal(new_object)
        new_object._state_bits = self._state_bits

        return new_object

    def __repr__(self):
        return "MineSweeperState[" + super(MineSweeperStateObject, self).__repr__() + ", state_bits=" + str(self._state_bits) + "\"]"

File: server/piwars-common/src/main/java/org/ah/piwars/virtualrover/game/MineSweeperStateObject.java
package org.ah.piwars.virtualrover.game;

import org.ah.themvsus.engine.common.game.GameObject;
import org.ah.themvsus.engine.common.game.GameObjectFactory;
import org.ah.themvsus.engine.common.game.GameObjectType;
import org.ah.themvsus.engine.common.transfer.Serializer;

public class MineSweeperStateObject extends GameObject {

    private int stateBits = 0x0;

    public MineSweeperStateObject(GameObjectFactory factory, int id) {
        super(factory, id);
        nonInteractive = true;
    }

    @Override
    public GameObjectType getType() { return PiWarsGameTypeObject.MineSweeperStateObject; }

    public int getStateBits() {
        return stateBits;
    }

    public void setStateBits(int stateBits) {
        changed = changed || this.stateBits != stateBits;
        this.stateBits = stateBits;
    }

    @Override
    public void serialize(boolean full, Serializer serializer) {
        super.serialize(full, serializer);

        serializer.serializeUnsignedShort(stateBits);
    }

    @Override
    public void deserialize(boolean full, Serializer serializer) {
        super.deserialize(full, serializer);

        int stateBits = serializer.deserializeUnsignedShort();
        setStateBits(stateBits);
    }

    @Override
    public int size(boolean full) {
        return super.size(full) + 2;
    }

    @Override
    protected GameObject copyInt(GameObject newObject) {
        super.copyInt(newObject);
        MineSweeperStateObject mineSweeperStateObject = (MineSweeperStateObject)newObject;
        mineSweeperStateObject.stateBits = stateBits;
        return mineSweeperStateObject;
    }

    @Override
    public String toString() {
        return "GameMessage" + (changed ? "*[" : "[") + id + ", stateBits=" + Integer.toHexString(stateBits) + "\"]";
    }
}
